### PR TITLE
Initial support for zone control

### PR DESCRIPTION
**Pull request by jwiese**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -224,7 +282,7 @@ Executing custom commands
 -------------------------
 
 For experimenting it can be useful to execute arbitrary commands against the endpoint.
-You can access the available methods by calling ``songpal list-all``.
+You can access the available methods by calling ``songpal list_all``.
```
Comment (rytilahti) [inline README.rst]: Please revert this.

File: songpal/containers.py [hunk 2]
```diff
@@ -263,7 +263,7 @@ def is_muted(self):
         return self.mute == "on"
 
     def __str__(self):
-        s = "Volume: %s/%s" % (self.volume, self.maxVolume)
+        s = "Zone %s Volume: %s/%s" % (self.output[self.output.rfind("=")+1:], self.volume, self.maxVolume)
```
Comment (rytilahti) [inline songpal/containers.py]: This feels potentially error-prone when it comes to devices with no zones, doesn't it?

File: songpal/containers.py [hunk 3]
```diff
@@ -340,9 +374,10 @@ def __str__(self):
             s += " (active)"
         return s
 
-    async def activate(self):
+    async def activate(self, output: Zone = None):
```
Comment (rytilahti) [inline songpal/containers.py]: Remove spaces around `=`.

File: songpal/containers.py [hunk 4]
```diff
@@ -340,9 +374,10 @@ def __str__(self):
...
         """Activate this input."""
-        return await self.services["avContent"]["setPlayContent"](uri=self.uri)
+        return await self.services["avContent"]["setPlayContent"](uri=self.uri, output=output.uri)
```
Comment (rytilahti) [inline songpal/containers.py]: I have not yet tested this in practice, but I would like to make sure this also works with devices without zones before this is merged.

File: songpal/discovery.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import logging
-from async_upnp_client.discovery import async_discover
+from async_upnp_client.search import async_search
```
Comment (rytilahti) [inline songpal/discovery.py]: Hmm, I thought this got already merged?

File: songpal/main.py [hunk 6]
```diff
@@ -227,20 +227,33 @@ def print_settings(settings, depth=0):
...
-            await input.activate()
         except StopIteration:
             click.echo("Unable to find input %s" % input)
             return
+        zone = None
+        if output:
+            zones = await dev.get_zones()
+            try:
+                zone = next((x for x in zones if x.title == output))
+            except StopIteration:
```
Comment (rytilahti) [inline songpal/main.py]: This all feels very fragile; I suppose it is fine for now, but I think at some point (after/if I get ever time to do refactoring), such permanent properties will be cached by the main device class.

File: songpal/main.py [hunk 7]
```diff
@@ -252,6 +265,36 @@ def print_settings(settings, depth=0):
...
+@pass_dev
+@coro
+async def zone(dev: Device, zone, activate):
+    """Get and change outputs."""
+    zones = await dev.get_zones()
+    if zone:
+        try:
+            zone = next((x for x in zones if x.title == zone))
+            
+            power = 'inactive' if activate == "off" else 'active'
```
Comment (rytilahti) [inline songpal/main.py]: Please modify `activate` to take a boolean as input, and define the click.argument as such. There should be no need for such constructs, not even in the cli parts :-)

**Review by rytilahti — CHANGES_REQUESTED** [review]
Looks generally good, thanks! :100:  I added a couple of points which needs to be cleared before this can be merged, just for the sake of maintainability.

File: README.rst (hunk 1 above)
Comment (jwiese) [reply README.rst]: Sorry, thought that was gone already, should be fixed now

File: songpal/containers.py (hunk 2 above)
Comment (jwiese) [reply songpal/containers.py]: I was assuming that there was always at least one zone, but I guess there are maybe some devices where that isn't the case? 

Ultimately, there are a couple of places that feel a little brittle like this because, as you mentioned below, `Zone`s are not cached in the device, which they probably should be. I had thought about it, but seemed like a bigger change to the overall structure of the library.

For now, I just changed this to see if `self.output` contained a string that contains '=', but happy to do something more substantial.

File: songpal/containers.py (hunk 3 above)
Comment (jwiese) [reply songpal/containers.py]: Done

File: songpal/containers.py (hunk 4 above)
Comment (jwiese) [reply songpal/containers.py]: Agreed. I don't have a device that doesn't have zones, so hard to mimic what the behavior is there in my own testing

File: songpal/discovery.py (hunk 5 above)
Comment (jwiese) [reply songpal/discovery.py]: Yeah, sorry my git-fu is a little weak, but this is *hopefully* fixed now?

File: songpal/main.py (hunk 6 above)
Comment (jwiese) [reply songpal/main.py]: Agreed! See comment above on the `Volume` print string as well. I'd be happy to help with that, but didn't want to make such a big change to your repository without some input/checking

File: songpal/main.py (hunk 7 above)
Comment (jwiese) [reply songpal/main.py]: Okay, did that, although I can't say that `songpal zone 'Main Zone' true` feels especially natural compared to `songpal zone 'Main Zone' on`. Feels inconsistent in the context of `songpal power on`

**Comment by jwiese** [comment]
I guess one question here is if there are additional functions that will be required to integrate this smoothly into Home Assistant. I don't have experience contributing to Home Assistant yet, so I'm not sure how that is going to go. This maybe gets back to @rytilahti's question about whether some of the zone logic should be cached in the device.

File: README.rst [hunk 8]
```diff
@@ -158,14 +158,19 @@ Volume Control
...
-    $ songpal volume [<value>|mute|unmute]
+    $ songpal volume [-- output 'output title'] [<value>|mute|unmute]
 
     $ songpal volume 20
 
     $ songpal volume +5
 
     $ songpal volume -10
 
+    $ songpal volume --output 'Zone 2'
```
Comment (rytilahti) [inline README.rst:169]: Here's what happens when calling when no zones are there:
```
DEBUG:songpal.service:getVolumeInformation got called with args (({},)) kwargs ({})
DEBUG:songpal.method:got res: {'id': 6,
 'result': [[{'maxVolume': 50,
              'minVolume': 0,
              'mute': 'off',
              'output': '',
              'step': 1,
              'volume': 20}]]}
Using output: Zone 2
DEBUG:songpal.service:getCurrentExternalTerminalsStatus got called with args (()) kwargs ({})
DEBUG:songpal.method:got res: {'id': 7,
 'result': [[{'active': 'inactive',
              'connection': 'unknown',
              'iconUrl': '',
              'label': '',
              'meta': 'meta:tv',
              'title': 'TV',
              'uri': 'extInput:tv'},
             {'active': 'inactive',
              'connection': 'unknown',
              'iconUrl': '',
              'label': '',
              'meta': 'meta:hdmi',
              'title': 'HDMI1',
              'uri': 'extInput:hdmi?port=1'},
             {'active': 'inactive',
              'connection': 'unknown',
              'iconUrl': '',
              'label': '',
              'meta': 'meta:hdmi',
              'title': 'HDMI2',
              'uri': 'extInput:hdmi?port=2'},
             {'active': 'active',
              'connection': 'unknown',
              'iconUrl': '',
              'label': '',
              'meta': 'meta:hdmi',
              'title': 'HDMI3',
              'uri': 'extInput:hdmi?port=3'},
             {'active': 'inactive',
              'connection': 'unknown',
              'iconUrl': '',
              'label': '',
              'meta': 'meta:btaudio',
              'title': 'Bluetooth Audio',
              'uri': 'extInput:btAudio'},
             {'active': 'inactive',
              'connection': 'unknown',
              'iconUrl': '',
              'label': '',
              'meta': 'meta:linemini',
              'title': 'Analog',
              'uri': 'extInput:line'}]]}
Traceback (most recent call last):
  File "/home/hass/home-assistant/libs/python-sony/songpal/main.py", line 359, in volume
    output_uri = next ((x.uri for x in await dev.get_zones() if x.title == output))
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/hass/venv/bin/songpal", line 11, in <module>
    load_entry_point('python-songpal', 'console_scripts', 'songpal')()
  File "/home/hass/venv/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/hass/venv/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/hass/venv/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/hass/venv/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/hass/venv/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/hass/venv/lib/python3.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/home/hass/venv/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/hass/home-assistant/libs/python-sony/songpal/main.py", line 35, in wrapper
    return loop.run_until_complete(f(*args, **kwargs))
  File "/usr/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
RuntimeError: coroutine raised StopIteration
```

File: songpal/containers.py (hunk 4 above)
Comment (rytilahti) [reply songpal/containers.py]: This indeed breaks when there is no zone available:
```
  File "/home/hass/home-assistant/libs/python-sony/songpal/containers.py", line 389, in activate
    return await self.services["avContent"]["setPlayContent"](uri=self.uri, output=output.uri)
AttributeError: 'NoneType' object has no attribute 'uri'
```

I think the easiest fix is to check for Noneness and pass `""` as output, then it works:
```
if output is None:
    output = ""
<call>
```

File: songpal/main.py [hunk 9]
```diff
@@ -250,6 +263,35 @@ def print_settings(settings, depth=0):
...
+async def zone(dev: Device, zone, activate):
+    """Get and change outputs."""
+    zones = await dev.get_zones()
+    if zone:
+        try:
+            zone = next((x for x in zones if x.title == zone))
+            
+            action = "Activating" if activate else "Deactivating"
+
+            click.echo("%s %s" % (action, zone))
```
Comment (rytilahti) [inline songpal/main.py]: Same here as above, make get_zones() to raise an error if no zones is available -> error handling comes automatically. And if zone name is defined, using the `get_zone(name)` should work, I suppose, making it necessary only to have the loop for printing out all the zones.

File: README.rst [hunk 10]
```diff
@@ -180,22 +185,75 @@ such as night mode or adjusting the subwoofer volume.
...
+        - extOutput:zone?zone=2
+        - extOutput:zone?zone=3
+        - extOutput:zone?zone=4
+      * VIDEO 2 (uri: extInput:video?port=2)
+        - extOutput:zone?zone=1
+        - extOutput:zone?zone=4
       * TV (uri: extInput:tv)
-      * HDMI1 (uri: extInput:hdmi?port=1)
-      * HDMI2 (uri: extInput:hdmi?port=2) (active)
-      * HDMI3 (uri: extInput:hdmi?port=3)
```
Comment (rytilahti) [inline README.rst:194]: Maybe keep these as an example when there is no zones available.

File: devinfos/STR-DN1060.json [hunk 11]
```diff
@@ -2502,4 +2502,4 @@
       "bssid" : null,
       "bleID" : null
    }
-}
+}
```
Comment (rytilahti) [inline devinfos/STR-DN1060.json]: Can you revert this change? (removal of end of file)

File: songpal/main.py (hunk 6 above)
Comment (rytilahti) [reply songpal/main.py]: After some thought, how about adding a helper called `get_zone(name)` which returns the zone if available, or raises an exception (SongpalException) when not?

The same should be done for `activate(zone)` when the input is not valid.

File: songpal/containers.py [hunk 12]
```diff
@@ -314,18 +317,59 @@ def __str__(self):
...
+
+    This implements __bool__() for easy checking if the device is turned on or not.
+    """
+    make = classmethod(make)
+
+    def _convert_title(x):
+        return x.strip()
+
+    def _convert_is_active(x):
+        return True if x == "active" else False
```
Comment (rytilahti) [inline songpal/containers.py]: There are already two instances of this very same code, could you please move `convert_is_active` to the top-level after `convert_to_bool` and get rid of the copies? Same for the title conversion, too!

**Review by rytilahti — COMMENTED** [review]
This is looking fine, I left a couple of comments and after those are done we can merge this!

File: songpal/main.py (hunk 7 above)
Comment (rytilahti) [reply songpal/main.py]: Ah, it indeed does not accept on or off. The proper fix would be to introduce a new type by extending the https://github.com/pallets/click/blob/master/click/types.py#L344 to check first if on or off is handled. This type should also be used directly in `power` command, but that can be left for another PR :-)

File: songpal/containers.py (hunk 2 above)
Comment (rytilahti) [reply songpal/containers.py]: I'll have to think about it a bit, but maybe the proper solution is to create a default zone (with output="") and replace Input with Zone altogether, as the functionality in both of those containers is very much the same?

**Comment by rytilahti** [comment]
> I guess one question here is if there are additional functions that will be required to integrate this smoothly into Home Assistant. I don't have experience contributing to Home Assistant yet, so I'm not sure how that is going to go. This maybe gets back to @rytilahti's question about whether some of the zone logic should be cached in the device.

See my comment above about "default" zone, I think that's the way to go and will make integrating with homeassistant easier. The current songpal platform has to be converted into a component (I have some code around, but haven't yet managed to finish the conversion), which shall then request the available zones from the device and initialize a `media_player` platform for each zone. This will need some added logic to deliver the notifications to correct instance per zone name.

File: songpal/containers.py (hunk 2 above)
Comment (rytilahti) [reply songpal/containers.py]: What kind of features each zone has? Source and volume control, but are there more? After this PR is merged, I'm going to do a heavy cleanup on the codebase to make it easier to maintain, and at that point it maybe makes sense to create a real zone class having methods related to controlling all zone-specific functionality.

Comment (jwiese) [reply songpal/containers.py]: - Input source
- Volume
- Activate (on/off)

These are different from Inputs (I think), because Inputs don't have a volume or a source associated with them (in fact, Inputs are the sources)

Inputs for Zones are a little funny because it seems that not all inputs can be activated for a zone, but this is listed on the inputs, not the zones, e.g.:

`
 {
      'active':'inactive',
      'connection':'unknown',
      'iconUrl':'',
      'label':'',
      'meta':'meta:bd-dvd',
      'outputs':[
         'extOutput:zone?zone=1',
         'extOutput:zone?zone=4'
      ],
      'title':'cast    ',
      'uri':'extInput:bd-dvd'
   },
`

File: songpal/containers.py (hunk 4 above)
Comment (jwiese) [reply songpal/containers.py]: Instead changing it to


```
output_uri = output.uri if output else ""
return await self.services["avContent"]["setPlayContent"](uri=self.uri, output=output_uri)

```

File: songpal/main.py (hunk 6 above)
Comment (jwiese) [reply songpal/main.py]: Added that helper to `Device` and replaced code for the `zone` `volume` and `input` CLI commands

File: README.rst (hunk 8 above)
Comment (jwiese) [reply README.rst:169]: The new call to `get_zone(name)` should fix this/handle gracefully

File: songpal/main.py (hunk 7 above)
Comment (jwiese) [reply songpal/main.py]: Done!

File: songpal/main.py (hunk 9 above)
Comment (jwiese) [reply songpal/main.py]: I believe this is resolved

File: devinfos/STR-DN1060.json (hunk 11 above)
Comment (jwiese) [reply devinfos/STR-DN1060.json]: done

File: README.rst (hunk 10 above)
Comment (jwiese) [reply README.rst:194]: Done

File: songpal/containers.py (hunk 12 above)
Comment (jwiese) [reply songpal/containers.py]: Done

**Comment by jwiese** [comment]
I think I got 'em all

**Review by rytilahti — APPROVED** [review]

**Comment by rytilahti** [comment]
Looks good, let's get it merged! :-) edit: I released 0.10 containing this, thanks again for your contribution! :1st_place_medal:
